### PR TITLE
Retry bulk request if all errors are configured retry error codes & Split large bulk requests

### LINKED ISSUE #4488: Retry Bulk Request If Bulk Request Item Fails
**Describe the feature:**
Follow up to https://github.com/JanusGraph/janusgraph/issues/4408. Since implementing that feature I've noticed bulk requests do not get retried. In particular I periodically will have `circuit_breaking_exception` error messages mentioned by the thrown IOException [here](https://github.com/JanusGraph/janusgraph/blob/487e10ca276678862fd8fb369d6d17188703ba67/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java#L429).

It appears the bulk API doesn't participate as expected with regards to `ResponseException`.

It appears it can "successfully" return a Bulk Response, that contains items within it that failed. This PR applies the retry logic originally added in https://github.com/JanusGraph/janusgraph/issues/4408 if all the failures within the bulk response set returned error codes that are configured for retry.

Circuit Breaker Exceptions in particular are coded [on the server side](https://github.com/elastic/elasticsearch/blob/db1c76bef514fa36935e14f51a9430a528f8c94e/server/src/main/java/org/elasticsearch/common/breaker/CircuitBreakingException.java#L67) as being another type of 429 Too Many Requests that would be appropriate for back pressure to be applied. Operationally some circuit breaker errors (like the parent circuit breaker) are OOM preventative measures to the calling client just saying your request _right now_ may cause an OOM, try again later.

**Describe a specific use case for the feature:**

This feature would allow bulk request item failures to be reactive to backpressure via error codes.

### PR DESCRIPTION
**Pull request by criminosis**
Closes #4488

Reused the retry logic from https://github.com/JanusGraph/janusgraph/pull/4409 to also apply to bulk requests if an inner bulk item request failed.

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there an issue associated with this PR? Is it referenced in the commit message?
- [X] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [X] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

**Comment by criminosis** [comment]
Had an idea for an improvement this morning to only retry the bulk items that returned an error. This is predicated upon the [Bulk API docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#bulk-api-response-body) stating the response's returned bulk item order is the same as the submission order.

> The bulk API’s response contains the individual results of each operation in the request, returned in the order submitted. The success or failure of an individual operation does not affect other operations in the request.

Wasn't sure if that would be preferable though, so I left it as an [unsquashed addition to the PR](https://github.com/JanusGraph/janusgraph/pull/4489/commits/5c2b132847d3d352e6a4af89475c8bf81e279c2a) if it'd be preferable to revert it. If it looks good happy to squash it down to true up the PR for merging.

**Comment by criminosis** [comment]
Added a test for bulk retry in 81b07ac68.

While I was working on that though I noticed on the ES docs state there is a size limit for bulk requests:

>There is no "correct" number of actions to perform in a single bulk request. Experiment with different settings to find the optimal size for your particular workload. Note that Elasticsearch limits the maximum size of a HTTP request to 100mb by default ...

I've bumped into this in my JG deployment, ES pushes back saying the Payload is too large. I didn't have a meaningful knob to turn from what I could tell so I opted to work around the issue by more aggressively splitting up calls on my JG client side. So since I was already working in the bulk request logic I implemented a simple request chunker.

If a bulk request has a payload that is too large for a single call, it'll split it up into multiple bulk request calls.

**Comment by criminosis** [comment]
@li-boxuan / @porunov might interest you since you reviewed the last one? I accidentally left the job half finished. Didn't realize bulk request errors were nested into the item level.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java [hunk 1]
```diff
@@ -416,17 +452,107 @@ public void bulkRequest(List<ElasticSearchMutation> requests, String ingestPipel
...
+        final List<Map<String, RestBulkResponse.RestBulkItemResponse>> bulkResponseItems,
+        final List<RequestBytes> submittedBulkRequestItems) {
+        final List<Triplet<Object, Integer, RequestBytes>> errors = new ArrayList<>(bulkResponseItems.size());
+        for (int itemIndex = 0; itemIndex < bulkResponseItems.size(); itemIndex++) {
+            Collection<RestBulkResponse.RestBulkItemResponse> bulkResponseItem = bulkResponseItems.get(itemIndex).values();
+            if (bulkResponseItem.size() > 1) {
+                throw new IllegalStateException("There should only be a single item per bulk reponse item entry");
+            }
+            RestBulkResponse.RestBulkItemResponse item = bulkResponseItem.iterator().next();
+            if (item.getError() != null && item.getStatus() != 404) {
```
Comment (li-boxuan) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:471]: (nit) can we avoid explicitly using `404`? ES library should have a constant variable for 404.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java [hunk 2]
```diff
@@ -416,17 +452,107 @@ public void bulkRequest(List<ElasticSearchMutation> requests, String ingestPipel
...
+                        //Only retry the bulk request if *all* the bulk response item error codes are retry error codes
+                        final Set<Integer> errorCodes = bulkItemsThatFailed.stream().map(Triplet::getValue1).collect(Collectors.toSet());
+                        if (retryCount < retryAttemptLimit && retryOnErrorCodes.containsAll(errorCodes)) {
+                            //Build up the next request batch, of only the failed mutations
+                            bulkRequestChunk = bulkItemsThatFailed.stream().map(Triplet::getValue2).collect(Collectors.toList());
+                            performRetryWait(retryCount);
+                            retryCount++;
+                        } else {
+                            final List<Object> errorItems = bulkItemsThatFailed.stream().map(Triplet::getValue0).collect(Collectors.toList());
+                            errorItems.forEach(error -> log.error("Failed to execute ES query: {}", error));
```
Comment (li-boxuan) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:548]: You don't need `{}` here
```suggestion
                            errorItems.forEach(error -> log.error("Failed to execute ES query:", error));
```

**Review by li-boxuan — COMMENTED** [review]
Thank you! This looks good to me in general! I wonder if you could add an integration test to `ElasticsearchJanusGraphIndexTest` that demonstrates this chunking behavior? You could set BULK_CHUNK_SIZE_LIMIT_BYTES to different numbers to test different scenario. I would imagine the following:

1) A very small `BULK_CHUNK_SIZE_LIMIT_BYTES` that would fail the request - demonstrating the `we've encountered an element we cannot send to Elasticsearch given the configured limit` scenario
2) A relatively small `BULK_CHUNK_SIZE_LIMIT_BYTES` that would make the request split into chunks. We probably have no way to enforce that in the test (we lack observability) but I guess we could at least use IDE debugger to verify that manually.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java (hunk 1 above)
Comment (criminosis) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java]: No objection from me it's just there because [it's what was there previously](https://github.com/JanusGraph/janusgraph/blob/e30b4d6ed84e243eddbc726e4304edd1091c3eb3/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java#L425).

Happy to change it. That's being compared against something from the Apache Http library, and it does provide status code constants like `HttpStatus.SC_NOT_FOUND`.

Switched in 09891809e

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java (hunk 2 above)
Comment (li-boxuan) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:549]: Oh damn sorry for that. I thought it was an exception

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java (hunk 1 above)
Comment (li-boxuan) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java]: Oh I didn’t notice that. Thx for polishing it!

**Comment by criminosis** [comment]
> A very small BULK_CHUNK_SIZE_LIMIT_BYTES that would fail the request - demonstrating the we've encountered an element we cannot send to Elasticsearch given the configured limit scenario

@li-boxuan I don't think it's possible to write such a test because JanusGraph doesn't treat such a scenario as a failure [in code](https://github.com/JanusGraph/janusgraph/blob/e30b4d6ed84e243eddbc726e4304edd1091c3eb3/janusgraph-core/src/main/java/org/janusgraph/graphdb/database/StandardJanusGraph.java#L1022-L1028) or in [documentation](https://docs.janusgraph.org/operations/recovery/#transaction-failure).

> If the primary persistence into the storage backend succeeds but secondary persistence into the indexing backends or the logging system fail, the transaction is still considered to be successful because the storage backend is the authoritative source of the graph.

So the response from the traversal comes back as a success. I guess I could assert that a query that uses the mixed index stored in ES fails to return the vertex because now that index is blind to the property...but that feels like a weird thing to test for, albeit it is congruent with intended behavior.

I've bumped into this asymmetric situation when I can query using an equality predicate (thus uses the data backend's composite index) returns the expected vertex, but because the write to the mixed index failed (successfully) if you use like a range predicate over the same value, it won't find the vertex.


> A relatively small BULK_CHUNK_SIZE_LIMIT_BYTES that would make the request split into chunks. We probably have no way to enforce that in the test (we lack observability) but I guess we could at least use IDE debugger to verify that manually.

Took a stab at this in af88b6a0a, set it as `@Disabled` so it doesn't contribute to the GH execution time given it has no assertions.

**Comment by criminosis** [comment]
@li-boxuan sounded like there wasn't anything objectionable with the additions, so I went ahead and squashed & signed off on the commit so the PR was back in a merge ready state again pending further feedback.

**Comment by li-boxuan** [comment]
> I guess I could assert that a query that uses the mixed index stored in ES fails to return the vertex because now that index is blind to the property...

A much more cumbersome is to read the transaction recovery log... but I won't set this as acceptance criteria.

I do think what you mentioned here seems promising. Doesn't seem weird to me :) we probably already have tests doing something similar.

**Comment by criminosis** [comment]
@li-boxuan wrote `writingAnItemLargerThanPermittedChunkLimitFails` as requested to assert the expected data asymmetry between the composite and mixed index if an item exceeds the permitted bulk chunk limit.  

An alternative idea I had was rather than abort the bulk write at the time that is observed would be to maybe still write the other elements that aren't in excess of the limit. Get in at least what _could_ go in rather than have that data be lost too.

In other words, given a batch of mutations of an exceptionally large mutation followed by some number of small ones, fail the first, but don't throw an exception until later.

**Comment by li-boxuan** [comment]
> In other words, given a batch of mutations of an exceptionally large mutation followed by some number of small ones, fail the first, but don't throw an exception until later.

i like that idea! Do you wanna open a follow-up PR for that 😉? Maybe a config to control ;)

**Comment by criminosis** [comment]
Yeah, that'd probably be best done as a followup PR.

**Review by li-boxuan — APPROVED** [review]
Thank you for continuously investing in this field!

**Comment by criminosis** [comment]
Happy to @li-boxuan. JanusGraph has been a critical piece to what I'm working on, hence why I've been bumping into these sharp edges 😅 .

Looking forward to getting a nightly build into my environment and put this change through its paces after whenever @porunov has time to give their review.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java [hunk 3]
```diff
@@ -416,17 +453,107 @@ public void bulkRequest(List<ElasticSearchMutation> requests, String ingestPipel
             APPEND_OP.apply(builder).append("refresh=").append(bulkRefresh);
         }
         builder.insert(0, REQUEST_SEPARATOR + "_bulk");
```
Comment (porunov) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:455]: Not related to your PR, but it seems it makes sense to move this line up (to be the next line after `StringBuilder` is created) to avoid array shifting operation. I.e. Just:
```
final StringBuilder builder = new StringBuilder();
builder.append(REQUEST_SEPARATOR).append("_bulk");
```

**Review by porunov — APPROVED** [review]
Thank you @criminosis for this great contribution and sorry for the late review.
This PR looks great to me. I have added small nitpicks, but I think it's OK to fix them in follow up PRs.

**Comment by criminosis** [comment]
@porunov I'd be happy to do this in the follow-up you mentioned. There's already a tentative PR in mind that I was talking about with @li-boxuan above.

Since you mentioned them as a follow-up is this PR valid for merging? I saw you tagged it for 1.1.0 so wasn't sure if that meant it was going to hang out unmerged for a period of time since there's also a live 1.0.1 milestone.

**Comment by porunov** [comment]
@criminosis Just merged this PR into ‘master’ branch. Thus, these changes will be available in 1.1.0 release and in the following “commit” releases. However, it won’t be backported to 1.0.1 release since it’s based on the feature which isn’t available in 1.0.0 release. If you want to port these changes to 1.0.1 release for any reason then feel free to open a PR targeting ‘v1.0’ branch with these changes and your previous changes.

**Comment by Allan-Clements** [comment]
Thanks @porunov! Having it 1.1.0 and the next commit release is perfectly fine. I was asking just as curiosity around merging logistics.

Excited to get this into my environment soon as I've been bumping into what prompted this PR fairly often.

I'll get rolling on the follow-ups. 👍

**Comment by criminosis** [comment]
Thanks @porunov! Having it 1.1.0 and the next commit release is perfectly fine. I was asking just as curiosity around merging logistics.

Excited to get this into my environment soon as I've been bumping into what prompted this PR fairly often.

I'll get rolling on the follow-ups. 👍

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java (hunk 3 above)
Comment (criminosis) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:455]: Looked into this this for my follow-up, but it seems like this would change behavior?

https://github.com/JanusGraph/janusgraph/blob/d255396dcddf5c19317a873ff1ce16e31f77a304/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java#L105

`APPEND_OP` either prefixes a `?` or a `&` if the builder is empty. So as currently written I believe it's producing paths like `/_bulk?pipeline=foo&refresh=true`.

If we put `builder.append(REQUEST_SEPARATOR).append("_bulk");` further up, I think it'd produce paths like `/_bulk&pipeline=foo&refresh=true` which wouldn't be properly formatting the query parameters anymore.

How does this look if we just use the string builder for the query parameters, and then just directly build the path? I originally used another StringBuilder for the path variable, but IntelliJ's linter suggested just doing String concatenations.

```
final StringBuilder bulkRequestQueryParameters = new StringBuilder();
        if (ingestPipeline != null) {
            APPEND_OP.apply(bulkRequestQueryParameters).append("pipeline=").append(ingestPipeline);
        }
        if (bulkRefreshEnabled) {
            APPEND_OP.apply(bulkRequestQueryParameters).append("refresh=").append(bulkRefresh);
        }
        final String bulkRequestPath = REQUEST_SEPARATOR + "_bulk" + bulkRequestQueryParameters;
        return Pair.with(bulkRequestPath, outputStream.toByteArray());
```

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java [hunk 4]
```diff
@@ -324,6 +324,11 @@ public class ElasticSearchIndex implements IndexProvider {
             "Comma separated list of Elasticsearch REST client ResponseException error codes to retry. " +
                 "E.g. \"408,429\"", ConfigOption.Type.LOCAL, String[].class, new String[0]);
 
+    public static final ConfigOption<Integer> BULK_CHUNK_SIZE_LIMIT_BYTES =
+        new ConfigOption<>(ELASTICSEARCH_NS, "bulk-chunk-size-limit-bytes",
+            "The total size limit in bytes of a bulk request. Mutation batches in excess of this limit will be " +
+                "chunked to this size.", ConfigOption.Type.LOCAL, Integer.class, 100_000_000);
+
```
[the remark this answers is not in the document]
Comment (criminosis) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java:331]:
```
 public static final ConfigOption<Integer> BULK_CHUNK_SIZE_LIMIT_BYTES =
        new ConfigOption<>(ELASTICSEARCH_NS, "bulk-chunk-size-limit-bytes",
            "The total size limit in bytes of a bulk request. Mutation batches in excess of this limit will be " +
                "chunked to this size. Ensure that this limit is always less than or equal to the configured limit of " +
                "`http.max_content_length` on the Elasticsearch servers. For more information, refer to the " +
                "[Elasticsearch documentation](https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-network.html).",
            ConfigOption.Type.LOCAL, Integer.class, 100_000_000);
```

Got it ready for the follow-up👍

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java (hunk 3 above)
Comment (porunov) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:455]: Oh. I missed that! I'm OK with leaving it as it is right now or updating it to the way you are suggesting. Both ways are OK by me.
btw. Usually `final` String concatenation is directly compiled to reuse `StringBuilder` by all modern Java compilers, so I believe that's why IntelliJ's suggested just using String concatenation here.
Anyways, the shift is probably too small to actually produce any meaningful performance differences, so I'm OK with either way as you prefer.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (porunov) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java:331]: Thanks! Would you be able to also add here or somewhere in the documentation description of what happens when part of the request produces a chunk which is grater than the one specified here?
I think this information will let users know that it's dangerous to reduce this chunk size to small sizes.

[the remark this answers is not in the document]
Comment (criminosis) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java:331]: Added `If a single bulk item exceeds this limit an exception will be thrown after the smaller bulk items are submitted.`  to the follow-up PR 👍 .

Figured we can continue the conversation of any further tweaking over there.
